lgc/disassembler/Disassembler.cpp

```diff
@@ -0,0 +1,506 @@
+  m_objFile.reset(cast<ELFObjectFileBase>(expectedObjFile.get().release()));
+
+  // Figure out the target triple from the object file, and get features.
+  Triple triple = m_objFile->makeTriple();
+  SubtargetFeatures features = m_objFile->getFeatures();
+
+  // Get the target specific parser.
+  std::string error;
+  m_tripleName = triple.getTriple();
+  m_target = TargetRegistry::lookupTarget(m_tripleName, error);
+  if (!m_target)
+    report_fatal_error(m_objFile->getFileName() + ": '" + m_tripleName + "': " + error);
+
+  // Get the CPU name.
+  Optional<StringRef> mcpu = m_objFile->tryGetCPUName();
+  if (!mcpu)
+    report_fatal_error(m_objFile->getFileName() + ": Cannot get CPU name");
+
+  // Output the required llvm-mc command as a comment.
+  m_ostream << "// llvm-mc -triple=" << m_tripleName << " -mcpu=" << mcpu << "\n";
+
+  // Set up other objects required for disassembly.
+  std::unique_ptr<MCRegisterInfo> regInfo(m_target->createMCRegInfo(m_tripleName));
+  if (!regInfo)
+    report_fatal_error(m_data.getBufferIdentifier() + ": No register info for target");
+  std::unique_ptr<MCAsmInfo> asmInfo(m_target->createMCAsmInfo(*regInfo, m_tripleName, MCTargetOptions()));
+  if (!asmInfo)
+    report_fatal_error(m_data.getBufferIdentifier() + ": No assembly info for target");
+  m_subtargetInfo.reset(m_target->createMCSubtargetInfo(m_tripleName, *mcpu, features.getString()));
+  if (!m_subtargetInfo)
+    report_fatal_error(m_data.getBufferIdentifier() + ": No subtarget info for target");
+  std::unique_ptr<MCInstrInfo> instrInfo(m_target->createMCInstrInfo());
+  if (!instrInfo)
+    report_fatal_error(m_data.getBufferIdentifier() + ": No instruction info for target");
+
+  MCObjectFileInfo objFileInfo;
+  MCContext context(asmInfo.get(), regInfo.get(), &objFileInfo);
+  m_context = &context;
+  objFileInfo.InitMCObjectFileInfo(triple, false, context);
+
+  m_instDisassembler.reset(m_target->createMCDisassembler(*m_subtargetInfo, *m_context));
+  if (!m_instDisassembler)
+    report_fatal_error(m_data.getBufferIdentifier() + ": No disassembler for target");
+  MCInstPrinter *instPrinter(
+      m_target->createMCInstPrinter(triple, asmInfo->getAssemblerDialect(), *asmInfo, *instrInfo, *regInfo));
+  if (!instPrinter)
+    report_fatal_error(m_data.getBufferIdentifier() + ": No instruction printer for target");
+
+  std::unique_ptr<formatted_raw_ostream> fostream(new formatted_raw_ostream(m_ostream));
+  m_streamer.reset(
+      m_target->createAsmStreamer(*m_context, std::move(fostream), true, false, instPrinter, nullptr, nullptr, false));
+
+  // Process each section.
+  for (ELFSectionRef sectionRef : m_objFile->sections())
+    processSection(sectionRef);
+}
+
+// =====================================================================================================================
+// Disassemble one section.
+void ObjDisassembler::processSection(ELFSectionRef sectionRef) {
```

Comment:
This function is very long. Would it be possible to split it into  a few smaller helper functions?